rustyms-py/src/lib.rs

```diff
@@ -0,0 +1,1262 @@
+    ///
+    /// Returns
+    /// -------
+    /// MolecularFormula
+    ///
+    fn formula_all(&self) -> MolecularFormula {
+        MolecularFormula(self.0.formula_all().unwrap())
+    }
+}
+
+/// A peptide with all data as provided by ProForma 2.0.
+///
+/// Parameters
+/// ----------
+/// proforma : str
+///     The ProForma string.
+///
+#[pyclass]
+#[derive(Clone)]
+pub struct LinearPeptide(rustyms::LinearPeptide);
+
+#[pymethods]
+impl LinearPeptide {
+    /// Create a new peptide from a ProForma string.
+    #[new]
+    fn new(proforma: &str) -> Self {
+        LinearPeptide(
+            rustyms::ComplexPeptide::pro_forma(proforma)
+                .unwrap()
+                .assume_linear(),
+        )
+    }
+
+    fn __str__(&self) -> String {
+        self.0.to_string()
+    }
+
+    fn __repr__(&self) -> String {
+        format!("LinearPeptide({})", self.0.to_string())
+    }
+
+    fn __len__(&self) -> usize {
+        self.0.sequence.len()
+    }
+
+    /// Labile modifications, which will not be found in the actual spectrum.
+    ///
+    /// Returns
+    /// -------
+    /// list[Modification]
+    ///
+    #[getter]
+    fn labile(&self) -> Vec<Modification> {
+        self.0
+            .labile
+            .iter()
+            .map(|x| Modification(x.clone()))
+            .collect()
+    }
+
+    /// N-terminal modification.
+    ///
+    /// Returns
+    /// -------
+    /// Modification | None
+    ///
+    #[getter]
+    fn n_term(&self) -> Option<Modification> {
+        match &self.0.n_term {
+            Some(m) => Some(Modification(m.clone())),
+            None => None,
+        }
+    }
+
+    /// C-terminal modification.
+    ///
+    /// Returns
+    /// -------
+    /// Modification | None
+    ///
+    #[getter]
+    fn c_term(&self) -> Option<Modification> {
+        match &self.0.c_term {
+            Some(m) => Some(Modification(m.clone())),
+            None => None,
+        }
+    }
+
+    /// Sequence of the peptide including modifications.
+    ///
+    /// Returns
+    /// -------
+    /// list[SequenceElement]
+    ///
+    #[getter]
+    fn sequence(&self) -> Vec<SequenceElement> {
+        self.0
+            .sequence
+            .iter()
+            .map(|x| SequenceElement(x.clone()))
+            .collect()
+    }
+
+    /// For each ambiguous modification list all possible positions it can be placed on. Indexed by the ambiguous modification id.
+    ///
+    /// Returns
+    /// -------
+    /// list[list[int]]
+    ///
+    #[getter]
+    fn ambiguous_modifications(&self) -> Vec<Vec<usize>> {
+        self.0.ambiguous_modifications.clone()
+    }
+
+    /// Stripped sequence, meaning the sequence without any modifications.
+    ///
+    /// Returns
+    /// -------
+    /// str
+    ///
+    #[getter]
+    fn stripped_sequence(&self) -> String {
+        self.0.sequence.iter().map(|x| x.aminoacid.char()).collect()
+    }
+
+    // TODO: How to get the charge as an integer?
```

Comment:
`self.formulas().next().unwrap().charge()` This takes the first formula and get the charge out of that, it unwraps because it assumes there to be at least one formula (which is always the cases) and only takes the first as the only difference will be in the count of C/H due to B/Z not in the actual charge. But I did not provide any method like this on the Rust side, if you think this is a useful method to have regardless I will set it up on the rust side as well.

I did not set it up I think because I never really regard any peptide as intrinsically charged, I just take it as a sequence and overwrite the charges in the generate_fragments method.

Reply:
Thanks!

It's indeed a bit strange to assign a charge in a ProForma peptidoform sequence, as it is a property of a precursor, not of a peptidoform. Nevertheless, it was added out of convenience for bottom up MS. For instance, I deal a lot with peptide spectrum MS2 intensities, which are mostly defined by the combination peptide-modifications-charge. For me, it would make sense to also have direct access to the charge state on the rust side.

Reply:
Good I will think about the cleanest way to implement it on the Rust side then, then this function can become a direct port again.

Reply:
Could it be that the charge is currently not parsed at all when creating a peptide with `ComplexPeptide::pro_forma`?

Reply:
There is a known limitation in my parser where it fails to parse c term modifications followed by a charge state and a chimeric peptide. I will have to address that at some point. But if you see something else that is going awry please let me know then I can add some more tests and fix it.

Reply:
It works now. Before I tried to get the charge from `formula()` instead of `charge_carriers().formula()`.